### PR TITLE
Take high price

### PR DESCRIPTION
**Pull request by ith-harvey**
## Description

* Created a test that disproves the concerns raised in https://github.com/sherlock-audit/2023-09-ajna-judging/issues/30
* When in ERC721 pools when the auctionPrice is above the top bucket price there cannot be a non-integral amount of collateral in the auction (`lenderKick` has no effect on collateral in auction and `bucketTake` cannot be called). Thus the concerns raised in the issue are invalid.
* Tests shows the amount of collateral emitted from the `take` event are inaccurate, see -> https://github.com/ajna-finance/contracts/pull/1016/files#diff-9e58db5fd372c3977d7c20bf27ae170489e045025c509c718035284dc56e3d14R1270

## Tasks

- [ ] Changes to protocol contracts are covered by unit tests executed by CI.
- [ ] Protocol contract size limits have not been exceeded.
- [ ] Gas consumption for impacted transactions have been compared with the target branch, and nontrivial changes cited in the _Impact_ section above.
- [x] Scope labels have been assigned as appropriate.
- [ ] Invariant tests have been manually executed as appropriate for the nature of the change.

### CODE REVIEW

File: tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsHighTake.t.sol [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+
+       _mintAndApproveCollateralTokens(_borrower,  6);
+       _mintAndApproveCollateralTokens(_borrower2, 68);
+       _mintAndApproveCollateralTokens(_borrower3, 2);
+
+        // Lender adds Quote token accross 5 prices
+        _addInitialLiquidity({
+            from:   _lender,
+            amount: 2_000 * 1e18,
+            index:  1
```
Comment (mattcushman) [inline tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsHighTake.t.sol]: I would simpligy this a lot.  Get rid of all the excess adds below this point, but have lender deposit a billion quote token at this level (which should be a price close to 1 billion)

File: tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsHighTake.t.sol [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        _addInitialLiquidity({
+            from:   _lender,
+            amount: 30_000.0 * 1e18,
+            index:  _i9_91
+        });
+
+       // first borrower adds collateral token and borrows
+        uint256[] memory tokenIdsToAdd = new uint256[](2);
+        tokenIdsToAdd[0] = 1;
+        tokenIdsToAdd[1] = 3;
```
Comment (mattcushman) [inline tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsHighTake.t.sol]: only add 1 token as collateral for borrower

File: tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsHighTake.t.sol [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+        // borrower deposits two NFTs into the subset pool and borrows
+        _pledgeCollateral({
+            from:     _borrower,
+            borrower: _borrower,
+            tokenIds: tokenIdsToAdd
+        });
+        _borrow({
+            from:       _borrower,
+            amount:     70_000.0 * 1e18,
```
Comment (mattcushman) [inline tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsHighTake.t.sol]: borrower has bid 1 collateral.  if there is a billion QT in bucket with index 1, they should be able to borrow close to 1 billion.

File: tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsHighTake.t.sol [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+            indexLimit: _i9_91,
+            newLup:     946585655.625225796879504875 * 1e18
+        });
+
+        // second borrower deposits three NFTs into the subset pool and borrows
+        tokenIdsToAdd = new uint256[](12);
```
Comment (mattcushman) [inline tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsHighTake.t.sol]: remove this, it's just noise from unit test point of view

File: tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsHighTake.t.sol [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+                referencePrice:    0,
+                totalBondEscrowed: 0,
+                auctionPrice:      0,
+                debtInAuction:     0,
+                thresholdPrice:    0,
+                neutralPrice:      0
+            })
+        );
+
+        _kick({
```
Comment (mattcushman) [inline tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsHighTake.t.sol]: might need to have lender `lenderKick` instead of kick

Comment (ith-harvey) [reply tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsHighTake.t.sol]: created a separate example with lenderKick -> https://github.com/ajna-finance/contracts/pull/1016/files#diff-8e166c398ee17ec3f4ec42561fc631c7e1df4848ec3a20d482d29d6a90e35b31R354

File: tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsTake.t.sol [hunk 6]
```diff
@@ -839,4 +839,479 @@ contract ERC721PoolLiquidationsTakeTest is ERC721HelperContract {
...
+        // top bucket price
+        assertEq(1_004_968_987.606512354182109771 * 1e18, _priceAt(0));
+
+        _take({
+            from:            _lender,
+            borrower:        _borrower,
+            maxCollateral:   2 * 1e18,
+            bondChange:      10_897_869.739154233968795067 * 1e18,
+            givenAmount:     991_360_820.988112808697327591 * 1e18,
+            collateralTaken: 0.003699377769661894 * 1e18,
```
Comment (mattcushman) [inline tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsTake.t.sol]: I'm surprised by this collateralTake < 1.  I thought that for NFTs collateralTaken (in a take()) was always an integer?

Comment (mattcushman) [reply tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsTake.t.sol]: I'm guessing that this is just how it's reported, as the borrower's collateral is actually wiped out.

**Review by mattcushman — APPROVED** [review]
I think it's fine, although did ask one question about the values asserted in `take()`

File: tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsTake.t.sol (hunk 6 above)
Comment (ith-harvey) [reply tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsTake.t.sol]: separate PR lives here -> https://github.com/ajna-finance/contracts/pull/1026

Comment (EdNoepel) [reply tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsTake.t.sol]: Recommend merging `develop` back into this branch again, as #1026 should seemingly change line 960.

Comment (ith-harvey) [reply tests/forge/unit/ERC721Pool/ERC721PoolLiquidationsTake.t.sol]: > Recommend merging `develop` back into this branch again, as #1026 should seemingly change line 960.

Merged `develop` back in

**Review by prat-gpt — APPROVED** [review]
LGTM

**Review by EdNoepel — APPROVED** [review]
This nicely validates the event emit change.

**Pull request merged**